# Data Quest: Data Scientist in Python

# Guided Project 4: Exploring eBay Car Sales Data

## Introduction:

This project uses a dataset of used car sales from a classified section of the German eBay website: eBay Kleinanzeigen. We'll use the concepts learned from Data Quest: Data Scientist in Python `Part 2 - Data Analysis and Visualization`.
The main project component is Python's `pandas` library, which we will use to clean, analyze and display data.

Chapters:
* Cleaning Column Names
* Initial Exploration and Cleaning
* Exploring the Odometer and Price Columns
* Exploring the Date Columns
* Dealing with Incorrect Registration Year Data
* Exploring Price by Brand
* Storing Aggregate Data in a DataFrame
* Further Data Cleaning and Analysis Steps

### Code Explanation:

Before we can start cleaning and analyzing data, we need to import both the `pandas` and `NumPy` libraries and assign them the aliases of `pd` and `np` for easier access. Next we need to read our dataset `autos.csv` using the pandas function `read_csv` - we give it the encoding `Latin-1` since the default `UTF-8` caused an error.

In [1]:
import pandas as pd
import numpy as np

autos = pd.read_csv('autos.csv', encoding='Latin-1')

Here we use one of Jupyter Notebook's built-in features to inspect the dataset. `autos` automatically reads out the  five first and last rows without a `print` statement or `indexing`.

In [2]:
autos

dateCrawled                                               name  \
0      2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1      2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2      2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3      2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4      2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   
...                    ...                                                ...   
49995  2016-03-27 14:38:19   Audi_Q5_3.0_TDI_qu._S_tr.__Navi__Panorama__Xenon   
49996  2016-03-28 10:50:25  Opel_Astra_F_Cabrio_Bertone_Edition___TÜV_neu+...   
49997  2016-04-02 14:44:48                     Fiat_500_C_1.2_Dualogic_Lounge   
49998  2016-03-08 19:25:42                 Audi_A3_2.0_TDI_Sportback_Ambition   
49999  2016-03-14 00:42:12                                Opel_Vectra_1.6_16V   

       seller offerType    price   abtest vehicleType  yearOfRegistration  \
0      privat   Angebot   $5,000  control         bus                2004   
1      privat   Angebot   $8,500  control   limousine                1997   
2      privat   Angebot   $8,990     test   limousine                2009   
3      privat   Angebot   $4,350  control  kleinwagen                2007   
4      privat   Angebot   $1,350     test       kombi                2003   
...       ...       ...      ...      ...         ...                 ...   
49995  privat   Angebot  $24,900  control   limousine                2011   
49996  privat   Angebot   $1,980  control      cabrio                1996   
49997  privat   Angebot  $13,200     test      cabrio                2014   
49998  privat   Angebot  $22,900  control       kombi                2013   
49999  privat   Angebot   $1,250  control   limousine                1996   

         gearbox  powerPS   model   odometer  monthOfRegistration fuelType  \
0        manuell      158  andere  150,000km                    3      lpg   
1      automatik      286     7er  150,000km                    6   benzin   
2        manuell      102    golf   70,000km                    7   benzin   
3      automatik       71  fortwo   70,000km                    6   benzin   
4        manuell        0   focus  150,000km                    7   benzin   
...          ...      ...     ...        ...                  ...      ...   
49995  automatik      239      q5  100,000km                    1   diesel   
49996    manuell       75   astra  150,000km                    5   benzin   
49997  automatik       69     500    5,000km                   11   benzin   
49998    manuell      150      a3   40,000km                   11   diesel   
49999    manuell      101  vectra  150,000km                    1   benzin   

            brand notRepairedDamage          dateCreated  nrOfPictures  \
0         peugeot              nein  2016-03-26 00:00:00             0   
1             bmw              nein  2016-04-04 00:00:00             0   
2      volkswagen              nein  2016-03-26 00:00:00             0   
3           smart              nein  2016-03-12 00:00:00             0   
4            ford              nein  2016-04-01 00:00:00             0   
...           ...               ...                  ...           ...   
49995        audi              nein  2016-03-27 00:00:00             0   
49996        opel              nein  2016-03-28 00:00:00             0   
49997        fiat              nein  2016-04-02 00:00:00             0   
49998        audi              nein  2016-03-08 00:00:00             0   
49999        opel              nein  2016-03-13 00:00:00             0   

       postalCode             lastSeen  
0           79588  2016-04-06 06:45:54  
1           71034  2016-04-06 14:45:08  
2           35394  2016-04-06 20:15:37  
3           33729  2016-03-15 03:16:28  
4           39218  2016-04-01 14:38:50  
...           ...                  ...  
4999

We print out information about our dataset using the pandas function - `DataFrame/Series.info()`, which gives us an idea of the different datatypes (`dtypes`) the dataset contains. 

We find that there are only two different types: `object` (strings) and `int64` (integers).

In [3]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 20 columns):
 #   Column               Non-Null Count  Dtype 
---  ------               --------------  ----- 
 0   dateCrawled          50000 non-null  object
 1   name                 50000 non-null  object
 2   seller               50000 non-null  object
 3   offerType            50000 non-null  object
 4   price                50000 non-null  object
 5   abtest               50000 non-null  object
 6   vehicleType          44905 non-null  object
 7   yearOfRegistration   50000 non-null  int64 
 8   gearbox              47320 non-null  object
 9   powerPS              50000 non-null  int64 
 10  model                47242 non-null  object
 11  odometer             50000 non-null  object
 12  monthOfRegistration  50000 non-null  int64 
 13  fuelType             45518 non-null  object
 14  brand                50000 non-null  object
 15  notRepairedDamage    40171 non-null  object
 16  date

We print out the column index names:

In [4]:
print(autos.columns)

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'odometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')


## Renaming Columns:

We want to rename some columns for enhanced readability and understanding.

### Code Explanation:

Function
1. We create a function `column_names()` taking one argument - `col`.
2. We replace the current column names with improved names using the `replace()` method - before returning the improved column names, we use the `lower()` method, to ensure that all names are lowercase letters.

Cleaned list
1. We create an empty list called `cleaned_columns` and start a `for` loop over the column names.
2. We use our function to assign each name to our variabel - `cleaned`
3. we append the newly cleaned names to our empty list, then passing the newly cleaned list back to `autos.columns` and printing it out to see the result.

In [7]:
def column_names(col):
    col = col.replace("yearOfRegistration", "reg_year")
    col = col.replace("monthOfRegistration", "reg_month")
    col = col.replace("notRepairedDamage", "unrepair_dmg")
    col = col.replace("dateCreated", "ad_created")
    col = col.replace("dateCrawled", "date_listing")
    col = col.replace("lastSeen", "last_seen")
    col = col.replace("nrOfPictures", "nrs_of_pictures")
    col = col.replace("postalCode", "postal_code")
    col = col.replace("fuelType", "fuel_type")
    col = col.replace("vehicleType", "vehicle_type")
    col = col.lower()
    return col

cleaned_columns = []

for c in autos.columns:
    cleaned = column_names(c)
    cleaned_columns.append(cleaned)

autos.columns = cleaned_columns

print(autos.columns)
print('\n')
print(autos.head())

Index(['date_listing', 'name', 'seller', 'offertype', 'price', 'abtest',
       'vehicle_type', 'reg_year', 'gearbox', 'powerps', 'model', 'odometer',
       'reg_month', 'fuel_type', 'brand', 'unrepair_dmg', 'ad_created',
       'nrs_of_pictures', 'postal_code', 'last_seen'],
      dtype='object')


          date_listing                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offertype   price   abtest vehicle_type  reg_year    gearbox  \
0  privat   Angebot  $5,000  control          bus      2004    manuell   
1  privat   Angebot  $8,500  control    limousine      1997  automatik   
2  privat   A

Here we print out the first five rows using the `DataFrame/Series.head()` function, which takes the agrument of `5` by default. 

We can choose the number of rows by writing `.head(number of rows)`.

In [8]:
autos.head()

date_listing                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller offertype   price   abtest vehicle_type  reg_year    gearbox  \
0  privat   Angebot  $5,000  control          bus      2004    manuell   
1  privat   Angebot  $8,500  control    limousine      1997  automatik   
2  privat   Angebot  $8,990     test    limousine      2009    manuell   
3  privat   Angebot  $4,350  control   kleinwagen      2007  automatik   
4  privat   Angebot  $1,350     test        kombi      2003    manuell   

   powerps   model   odometer  reg_month fuel_type       brand unrepair_dmg  \
0      158  andere  150,000km          3       lpg     peugeot         nein   
1      286     7er  150,000km          6    benzin         bmw         nein   
2      102    golf   70,000km          7    benzin  volkswagen         nein   
3       71  fortwo   70,000km          6    benzin       smart         nein   
4        0   focus  150,000km          7    benzin        ford         nein   

            ad_created  nrs_of_pictures  postal_code            last_seen  
0  2016-03-26 00:00:00                0        79588  2016-04-06 06:45:54  
1  2016-04-04 00:00:00                0        71034  2016-04-06 14:45:08  
2  2016-03-26 00:00:00                0        35394  2016-04-06 20:15:37  
3  2016-03-12 00:00:00                0        33729  2016-03-15 03:16:28  
4  2016-04-01 00:00:00                0        39218  2016-04-01 14:38:50

## Initial Exploring and Cleaning:

Before we can begin to analyze the data, we need to explor the dataset for any data that needs cleaning, initially we are looking for:

* Text columns where all or almost all the values are the same - they can be dropped as they aren't useful for the analysis.
* Numeric data stored as text strings - they can be cleaned and converted to `int/float` datatypes.

To start this process, we use the pandas function `.describe()`, which provides us with a statistical summary of our dataset - the parameter `include='all'` allows for both numeric and non-numeric column data descriptions.

In [9]:
autos.describe(include='all')

date_listing         name  seller offertype  price abtest  \
count                 50000        50000   50000     50000  50000  50000   
unique                48213        38754       2         2   2357      2   
top     2016-03-29 23:42:13  Ford_Fiesta  privat   Angebot     $0   test   
freq                      3           78   49999     49999   1421  25756   
mean                    NaN          NaN     NaN       NaN    NaN    NaN   
std                     NaN          NaN     NaN       NaN    NaN    NaN   
min                     NaN          NaN     NaN       NaN    NaN    NaN   
25%                     NaN          NaN     NaN       NaN    NaN    NaN   
50%                     NaN          NaN     NaN       NaN    NaN    NaN   
75%                     NaN          NaN     NaN       NaN    NaN    NaN   
max                     NaN          NaN     NaN       NaN    NaN    NaN   

       vehicle_type      reg_year  gearbox       powerps  model   odometer  \
count         44905  50000.000000    47320  50000.000000  47242      50000   
unique            8           NaN        2           NaN    245         13   
top       limousine           NaN  manuell           NaN   golf  150,000km   
freq          12859           NaN    36993           NaN   4024      32424   
mean            NaN   2005.073280      NaN    116.355920    NaN        NaN   
std             NaN    105.712813      NaN    209.216627    NaN        NaN   
min             NaN   1000.000000      NaN      0.000000    NaN        NaN   
25%             NaN   1999.000000      NaN     70.000000    NaN        NaN   
50%             NaN   2003.000000      NaN    105.000000    NaN        NaN   
75%             NaN   2008.000000      NaN    150.000000    NaN        NaN   
max             NaN   9999.000000      NaN  17700.000000    NaN        NaN   

           reg_month fuel_type       brand unrepair_dmg           ad_created  \
count   50000.000000     45518       50000        40171                50000   
unique           NaN         7          40            2                   76   
top              NaN    benzin  volkswagen         nein  2016-04-03 00:00:00   
freq             NaN     30107       10687        35232                 1946   
mean        5.723360       NaN         NaN          NaN                  NaN   
std         3.711984       NaN         NaN          NaN                  NaN   
min         0.000000       NaN         NaN          NaN                  NaN   
25%         3.000000       NaN         NaN          NaN                  NaN   
50%         6.000000       NaN         NaN          NaN                  NaN   
75%         9.000000       NaN         NaN          NaN                  NaN   
max        12.000000       NaN         NaN          NaN                  NaN   

        nrs_of_pictures   postal_code            last_seen  
count           50000.0  50000.000000                50000  
unique              NaN           NaN                39481  
top                 NaN           NaN  2016-04-07 06:17:27  
freq                NaN           NaN                    8  
mean                0.0  50813.627300                  NaN  
std                 0.0  25779.747957                  NaN  
min                 0.0   1067.000000                  NaN  
25%                 0.0  30451.000000                  NaN  
50%                 0.0  49577.000000                  NaN  
75%                 0.0  71540.000000                  NaN  
max                 0.0  99998.000000                  NaN

 `NaN` or `Not a Number` just means that the data is of non-numeric value.
 
## Exploring the Odomter and Price Columns:

After running the code above, we see that the `odometer` and `price` columns need cleaning - we can remove the `km`, `,`, and `$` - and turn them into numeric values (`int/float`).

First lets print them out, so we can compare the results:

### Before

In [12]:
autos['odometer'].head()

0    150,000km
1    150,000km
2     70,000km
3     70,000km
4    150,000km
Name: odometer, dtype: object

In [13]:
autos['price'].head()

0    $5,000
1    $8,500
2    $8,990
3    $4,350
4    $1,350
Name: price, dtype: object

Next, lets replace the unwanted strings and convert them to `integers` - we also rename the odometer to `odometer_km` to keep the code understandable for later:

### After

In [14]:
autos['odometer'] = autos['odometer'].str.replace(',', '').str.replace('km', '').astype(int)
autos.rename({'odometer': 'odometer_km'}, axis=1, inplace=True)

autos['odometer_km'].head()

0    150000
1    150000
2     70000
3     70000
4    150000
Name: odometer_km, dtype: int64

In [15]:
autos['price'] = autos['price'].str.replace(',', '').str.replace('$', '').astype(int)

autos['price'].head()

0    5000
1    8500
2    8990
3    4350
4    1350
Name: price, dtype: int64

Next we'll analyze the columns to see any unrealistic or irregular values, for this we use the `min/max()` function:

In [17]:
min_km = autos['odometer_km'].min()
max_km = autos['odometer_km'].max()

min_price = autos['price'].min()
max_price = autos['price'].max()

print('Maximum Odometer Distance: ', max_km, 'km\nMinimum Odomter Distance: ', min_km, 'km\n')
print('Maximum Price: ', max_price, '$\nMinimum Price: ', min_price, '$')

Maximum Odometer Distance:  150000 km
Minimum Odomter Distance:  5000 km

Maximum Price:  99999999 $
Minimum Price:  0 $


### Exploring the Odomter Column Further:

Lets explore the `odometer_km` column further to find any corrupt or incorrect data that can be removed.
We want to know `unique values` like: min/max/median/mean, and different value counts - for this we use these pandas functions:
`DataFrame/Series.shape()` - returns a tuple with information about number of rows and columns
`DataFrame/Series.describe()` - reutrns descriptive statistics - count, mean, standard deviation, min/max, and quartiles.
`DataFrame/Series.value_counts()` - returns a series that contains counts of unique values.

In [21]:
unique = autos['odometer_km'].unique().shape
des = autos['odometer_km'].describe()
val = autos['odometer_km'].value_counts().head(15)

print('Unique: ', unique , '\n')
print('--------------------------------\n')
print('Description: \n',des, '\n')
print('--------------------------------\n')
print('Value Counts: \n ', val, '\n')

Unique:  (13,) 

--------------------------------

Description: 
 count     50000.000000
mean     125732.700000
std       40042.211706
min        5000.000000
25%      125000.000000
50%      150000.000000
75%      150000.000000
max      150000.000000
Name: odometer_km, dtype: float64 

--------------------------------

Value Counts: 
  150000    32424
125000     5170
100000     2169
90000      1757
80000      1436
70000      1230
60000      1164
50000      1027
5000        967
40000       819
30000       789
20000       784
10000       264
Name: odometer_km, dtype: int64 



We note that there are `13` unique rows in the `odometer_km` column, and that most cars in the dataset a high odometer distance - we will leave this column as is, for now.

### Exploring the Price Column Further:

Now, let's do the exact same thing for the `price` column:

In [22]:
unique = autos['price'].unique().shape
des = autos['price'].describe()
val = autos['price'].value_counts().head(15)

print('Unique: ', unique , '\n')
print('--------------------------------\n')
print('Description: \n',des, '\n')
print('--------------------------------\n')
print('Value Counts: \n ', val, '\n')

Unique:  (2357,) 

--------------------------------

Description: 
 count    5.000000e+04
mean     9.840044e+03
std      4.811044e+05
min      0.000000e+00
25%      1.100000e+03
50%      2.950000e+03
75%      7.200000e+03
max      1.000000e+08
Name: price, dtype: float64 

--------------------------------

Value Counts: 
  0       1421
500      781
1500     734
2500     643
1000     639
1200     639
600      531
800      498
3500     498
2000     460
999      434
750      433
900      420
650      419
850      410
Name: price, dtype: int64 



We note an irregularity here. There are `1421` rows of cars where the price is `0` - this column needs to be investigated further.

We'll find out the highest and lowest prices by using the `min/max()` function:

In [25]:
max_price = autos['price'].value_counts().sort_index(ascending=False).head(15)
min_price = autos['price'].value_counts().sort_index(ascending=True).head(15)

print('Max Price Value Counts: \n', max_price)
print('\nMin Price Value Counts: \n', min_price)

Max Price Value Counts: 
 99999999    1
27322222    1
12345678    3
11111111    2
10000000    1
3890000     1
1300000     1
1234566     1
999999      2
999990      1
350000      1
345000      1
299000      1
295000      1
265000      1
Name: price, dtype: int64

Min Price Value Counts: 
 0     1421
1      156
2        3
3        1
5        2
8        1
9        1
10       7
11       2
12       3
13       2
14       1
15       2
17       3
18       1
Name: price, dtype: int64


Earlier we found that the maximum price for a car was `99,999,999$`, which is not very feasable. However, after inspecting the `Max Price Value Counts`, we find that there are only `14` cars with a price higher than `350,000$`, and since prices above 350k almost seem fake, we opt to remove listings with a price over that amount.

We also noted earlier that there are `1421` cars with a price of `0$`, and no one is giving away their cars. However, since eBay is an auction website, we opt to keep all prices from `1$` and up - since bidding can start at `1$`.

The code below is used to clean the `price` column for any prices higher than `350,000$` and lower than `1$`:

In [26]:
autos = autos[autos['price'].between(1, 350000)]
autos['price'].describe()

count     48565.000000
mean       5888.935591
std        9059.854754
min           1.000000
25%        1200.000000
50%        3000.000000
75%        7490.000000
max      350000.000000
Name: price, dtype: float64

## Exploring the Date Columns:

Next, we'll look at all the columns containing dates, these five are of interest:
* date_listing 
* last_seen
* ad_created
* reg_month
* reg_year


We need to convert the first three columns from strings to numeric values, but lets print them out first for easier comparision and information access:

In [27]:
columns = autos.iloc[:, [0, 16, 19]]
print(columns[:10])

          date_listing           ad_created            last_seen
0  2016-03-26 17:47:46  2016-03-26 00:00:00  2016-04-06 06:45:54
1  2016-04-04 13:38:56  2016-04-04 00:00:00  2016-04-06 14:45:08
2  2016-03-26 18:57:24  2016-03-26 00:00:00  2016-04-06 20:15:37
3  2016-03-12 16:58:10  2016-03-12 00:00:00  2016-03-15 03:16:28
4  2016-04-01 14:38:50  2016-04-01 00:00:00  2016-04-01 14:38:50
5  2016-03-21 13:47:45  2016-03-21 00:00:00  2016-04-06 09:45:21
6  2016-03-20 17:55:21  2016-03-20 00:00:00  2016-03-23 02:48:59
7  2016-03-16 18:55:19  2016-03-16 00:00:00  2016-04-07 03:17:32
8  2016-03-22 16:51:34  2016-03-22 00:00:00  2016-03-26 18:18:10
9  2016-03-16 13:47:02  2016-03-16 00:00:00  2016-04-06 10:46:35


In [28]:
listing = (autos['date_listing'].str[:10].value_counts(normalize=True, dropna=False).sort_index())
print(listing)

2016-03-05    0.025327
2016-03-06    0.014043
2016-03-07    0.036014
2016-03-08    0.033296
2016-03-09    0.033090
2016-03-10    0.032184
2016-03-11    0.032575
2016-03-12    0.036920
2016-03-13    0.015670
2016-03-14    0.036549
2016-03-15    0.034284
2016-03-16    0.029610
2016-03-17    0.031628
2016-03-18    0.012911
2016-03-19    0.034778
2016-03-20    0.037887
2016-03-21    0.037373
2016-03-22    0.032987
2016-03-23    0.032225
2016-03-24    0.029342
2016-03-25    0.031607
2016-03-26    0.032204
2016-03-27    0.031092
2016-03-28    0.034860
2016-03-29    0.034099
2016-03-30    0.033687
2016-03-31    0.031834
2016-04-01    0.033687
2016-04-02    0.035478
2016-04-03    0.038608
2016-04-04    0.036487
2016-04-05    0.013096
2016-04-06    0.003171
2016-04-07    0.001400
Name: date_listing, dtype: float64


In [29]:
ad_created = (autos['ad_created'].str[:10].value_counts(normalize=True, dropna=False).sort_index())
print(ad_created)

2015-06-11    0.000021
2015-08-10    0.000021
2015-09-09    0.000021
2015-11-10    0.000021
2015-12-05    0.000021
                ...   
2016-04-03    0.038855
2016-04-04    0.036858
2016-04-05    0.011819
2016-04-06    0.003253
2016-04-07    0.001256
Name: ad_created, Length: 76, dtype: float64


In [30]:
last_seen = (autos['last_seen'].str[:10].value_counts(normalize=True, dropna=False).sort_index())
print(last_seen)

2016-03-05    0.001071
2016-03-06    0.004324
2016-03-07    0.005395
2016-03-08    0.007413
2016-03-09    0.009595
2016-03-10    0.010666
2016-03-11    0.012375
2016-03-12    0.023783
2016-03-13    0.008895
2016-03-14    0.012602
2016-03-15    0.015876
2016-03-16    0.016452
2016-03-17    0.028086
2016-03-18    0.007351
2016-03-19    0.015834
2016-03-20    0.020653
2016-03-21    0.020632
2016-03-22    0.021373
2016-03-23    0.018532
2016-03-24    0.019767
2016-03-25    0.019211
2016-03-26    0.016802
2016-03-27    0.015649
2016-03-28    0.020859
2016-03-29    0.022341
2016-03-30    0.024771
2016-03-31    0.023783
2016-04-01    0.022794
2016-04-02    0.024915
2016-04-03    0.025203
2016-04-04    0.024483
2016-04-05    0.124761
2016-04-06    0.221806
2016-04-07    0.131947
Name: last_seen, dtype: float64


### Findings:

We see that similar dates from the `date_listing` column and the `ad_created` column have similar values - which makes sense, as advertisements are made for the new listings. The lower percentages of the `ad_created` column are most likely explained by completed car sales - meaning the ad won't be used anymore.

We also see a notable increase in percentages on the last three dates of the `last_seen` column.

## Dealing with Incorrect Registration Year Data:

Next, lets look the `reg_year` column to find any incorrect data:

In [34]:
print('Earliest Registration Year:', autos['reg_year'].min())
print('Latest Registration Year: ', autos['reg_year'].max())

Earliest Registration Year: 1000
Latest Registration Year:  9999


We find that the `min/max()` values for the `reg_year` column are incorrect, one is before the car was invented, and the other is 7976 years into the future.

So, we'll remove any year lower than `1886` (invention of the car) and `2016` (dataset origin date) to see how many percentages we are left with: 

In [35]:
count = (autos['reg_year'].between(1886, 2016).sum() / autos.shape[0])
print('Percentages of Cars Registered Between 1886 & 2016: ',round(count, 2), '%')

Percentages of Cars Registered Between 1886 & 2016:  0.96 %


After cleaning the dataset, we are left with `96%` of the cars, and we removed `4%` of imaginary cars. 

Next we want to know the 10 most frequent registration years:

In [36]:
autos = autos[autos["reg_year"].between(1886, 2016)]
year_count = autos["reg_year"].value_counts(normalize=True).head(10)
print(round(year_count, 3))
print(round(year_count.sum(),3))

2000    0.068
2005    0.063
1999    0.062
2004    0.058
2003    0.058
2006    0.057
2001    0.056
2002    0.053
1998    0.051
2007    0.049
Name: reg_year, dtype: float64
0.575


We find that `57.5%` of cars are registered in the last 25 years.

## Exploring Price by Brand:

Now, let's have a look at the prices for different car brands, for this we'll use loops to perform `aggregation` (combining multiple data points into a single value).

First we'll select the brands that holds `5 or more %` of the market share:

In [38]:
brands = autos["brand"].value_counts(normalize=True)
print(brands.head(20))

print('\n')

top_brands = brands[brands > 0.05].index
for brand in top_brands:
    print(brand)

volkswagen        0.211264
bmw               0.110045
opel              0.107581
mercedes_benz     0.096463
audi              0.086566
ford              0.069900
renault           0.047150
peugeot           0.029841
fiat              0.025642
seat              0.018273
skoda             0.016409
nissan            0.015274
mazda             0.015188
smart             0.014160
citroen           0.014010
toyota            0.012703
hyundai           0.010025
sonstige_autos    0.009811
volvo             0.009147
mini              0.008762
Name: brand, dtype: float64


volkswagen
bmw
opel
mercedes_benz
audi
ford


The top 5 brands are - Volkswagen, BMW, Mercedes Benz, Audi, and Ford.

This shows how German-made cars are dominate the market, having the four brands on the top 5, and those four being higher rated than Ford.

Now, lets find the average price for cars in the top 6:

In [40]:
avg_price = {}

for brand in top_brands:
    name = autos[autos['brand'] == brand]
    price = name['price'].mean()
    avg_price[brand] = int(price)
    
avg_price

{'volkswagen': 5402,
 'bmw': 8332,
 'opel': 2975,
 'mercedes_benz': 8628,
 'audi': 9336,
 'ford': 3749}

### Findings:

We can see that the highest average price is `9336` (Audi) and the lowest is `2975` (Opel), and that the most popular brand `Volkswagen`'s price (5402) is almost the mean/average between Audi and Opel which is: `6155.5`. We also notice that the brands - Audi, BMW, and Mercedes Benz tend to sell at higher prices. 

## Storing Aggregate Data in DataFrame:

Next we are going to compare the average mileage with the prices of the top 6 car brands.

We can use the same method we did for average price to calculate the average mileage:
### Average mileage:

In [41]:
avg_mile = {}

for brand in top_brands:
    name = autos[autos['brand'] == brand]
    mile = name['odometer_km'].mean()
    avg_mile[brand] = int(mile)
    
avg_mile

{'volkswagen': 128707,
 'bmw': 132572,
 'opel': 129310,
 'mercedes_benz': 130788,
 'audi': 129157,
 'ford': 124266}

### Average mileage & price together:

In [46]:
avg_mile = pd.Series(avg_mile).sort_values(ascending=True)
avg_price = pd.Series(avg_price).sort_values(ascending=True)

print('Average Mileage by Brand:\n\n ',avg_mile)
print('\nAverage Price by Brand:\n\n ',avg_price)

Average Mileage by Brand:

  ford             124266
volkswagen       128707
audi             129157
opel             129310
mercedes_benz    130788
bmw              132572
dtype: int64

Average Price by Brand:

  opel             2975
ford             3749
volkswagen       5402
bmw              8332
mercedes_benz    8628
audi             9336
dtype: int64


Now lets add both these `Series` to a `DataFrame` to see the top 6 car brands average price and mileage:

In [53]:
brand_detail = pd.DataFrame(avg_mile, columns = ['avg_mileage'])
brand_detail

avg_mileage
ford                124266
volkswagen          128707
audi                129157
opel                129310
mercedes_benz       130788
bmw                 132572

In [54]:
brand_detail['avg_price'] = avg_price
brand_detail

avg_mileage  avg_price
ford                124266       3749
volkswagen          128707       5402
audi                129157       9336
opel                129310       2975
mercedes_benz       130788       8628
bmw                 132572       8332

### Findings:

We found that the cars mileage isn't a huge factor when it comes to pricing. However, we found that the more expensive cars tended to have a higher mileage, which is suprising - `Opel` being the outlier.

## Further Data Cleaning and Analysis Steps:

### Data Cleaning:
We'll continue with cleaning up the data:
* Identifying categorical data that uses german words - translate them and map the value to their english counterparts
* Convert the dates to be uniform numeric data - `"2016-03-21"` becomes integer `20160321`.
* See if there are particular keywords in the name column that can be extracted as new columns.

### Translating German words to English:

The rows:
* gearbox
* fueltype
* unrep_damage
* vehicletype
* model, 
all have German words that need translating.

1. First we print out all the rows with the `.value_counts()` function.
2. Then we create a function called `translate()` to simplify the translations - the function takes three agruments - `df = autos dataframe`, `row_name = 'gearbox'`, and `translation = 'nein': 'no'` (dictionary for translations).
3. lastly we print out the first five rows using the pandas `.head()` method to verify our changes.

In [62]:
autos["gearbox"].value_counts()

manuell      34715
automatik     9856
Name: gearbox, dtype: int64

In [63]:
autos["fuel_type"].value_counts()

benzin     28540
diesel     14032
lpg          649
cng           71
hybrid        37
elektro       19
andere        15
Name: fuel_type, dtype: int64

In [64]:
autos["unrepair_dmg"].value_counts()

nein    33834
ja       4540
Name: unrepair_dmg, dtype: int64

In [65]:
autos["vehicle_type"].value_counts()

limousine     12598
kleinwagen    10585
kombi          8930
bus            4031
cabrio         3016
coupe          2462
suv            1965
andere          390
Name: vehicle_type, dtype: int64

Now, let's create a function to translate the German-words to English:

In [67]:
def translate(df, row_name, translation):
    df[row_name] = df[row_name].replace(translation)

gearbox = {'manuell': 'manual', 'automatik': 'automatic'}   
translate(autos, 'gearbox', gearbox)

fuel = {'benzin': 'gasoline', 'elektro': 'electric', 'andere': 'other'}
translate(autos, 'fuel_type', fuel)

unrep = {'nein': 'no', 'ja': 'yes'}
translate(autos, 'unrepair_dmg', unrep)

v_type = {'limousine': 'sedan', 'kleinwagen': 'small car', 'kombi': 'station wagon', 'cabrio': 'convertable', 'andere': 'other'}
translate(autos, 'vehicle_type', v_type)

model = {'andere': 'other'}
translate(autos, 'model', model)

o_type = {'Angebot': 'Offer/Bid'}
translate(autos, 'offertype', o_type)

autos.head()

date_listing                                               name  \
0  2016-03-26 17:47:46                   Peugeot_807_160_NAVTECH_ON_BOARD   
1  2016-04-04 13:38:56         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik   
2  2016-03-26 18:57:24                         Volkswagen_Golf_1.6_United   
3  2016-03-12 16:58:10  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...   
4  2016-04-01 14:38:50  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...   

   seller  offertype  price   abtest   vehicle_type  reg_year    gearbox  \
0  privat  Offer/Bid   5000  control            bus      2004     manual   
1  privat  Offer/Bid   8500  control          sedan      1997  automatic   
2  privat  Offer/Bid   8990     test          sedan      2009     manual   
3  privat  Offer/Bid   4350  control      small car      2007  automatic   
4  privat  Offer/Bid   1350     test  station wagon      2003     manual   

   powerps   model  odometer_km  reg_month fuel_type       brand unrepair_dmg  \
0      158   other       150000          3       lpg     peugeot           no   
1      286     7er       150000          6  gasoline         bmw           no   
2      102    golf        70000          7  gasoline  volkswagen           no   
3       71  fortwo        70000          6  gasoline       smart           no   
4        0   focus       150000          7  gasoline        ford           no   

            ad_created  nrs_of_pictures  postal_code            last_seen  
0  2016-03-26 00:00:00                0        79588  2016-04-06 06:45:54  
1  2016-04-04 00:00:00                0        71034  2016-04-06 14:45:08  
2  2016-03-26 00:00:00                0        35394  2016-04-06 20:15:37  
3  2016-03-12 00:00:00                0        33729  2016-03-15 03:16:28  
4  2016-04-01 00:00:00                0        39218  2016-04-01 14:38:50

### Converting Dates to Type - int:

1. First we extract the 10 first strings of the `listing_date` row - which is the dates in this syntax - `2016-03-21`.
2. Next we take the `listing_date` row and use the `pd.to_datetime` to convert the `autos` dataframe from strings to pandas `datetime64`.
3. Next we use the `dt.strftime()` method to format the datetime from `2016-03-26` (`%Y%m%d`) to `20160326`.
4. Then we convert the row values to type `int` using pandas method `.astype()`.
5. Lastly, we update the variable `listing_date` with the new date values and print them out to verify.

In [69]:
listings = (autos['date_listing'].str[:10])
autos['date_listing'] = pd.to_datetime(autos['date_listing'])
autos['date_listing'] = autos['date_listing'].dt.strftime('%Y%m%d')
autos['date_listing'].astype(int)
listings = autos['date_listing'][:10]
print(listings)

0    20160326
1    20160404
2    20160326
3    20160312
4    20160401
5    20160321
6    20160320
7    20160316
8    20160322
9    20160316
Name: date_listing, dtype: object


### Cleaning up the 'name' Column:

Here we remove the extra car name details and add a new column called `details` to store this information. We want to keep the first two strings in each car name - f.ex, `BMW 740i, Peugeot 807` etc. The extra information (`60_NAVTECH_ON_BOARD`) is not needed.

1. We first we split the string using `str.split()` and join the two strings together seperated by a space with `.str[:2].str.join(' ')`
2. Then we create a new column named `details` and do the same to every string after the two first.

In [70]:
name = autos['name'].str.split('_').str[:2].str.join(' ')
autos['detail'] = autos['name'].str.split('_').str[2:].str.join(' ')

autos.head()

date_listing                                               name  seller  \
0     20160326                   Peugeot_807_160_NAVTECH_ON_BOARD  privat   
1     20160404         BMW_740i_4_4_Liter_HAMANN_UMBAU_Mega_Optik  privat   
2     20160326                         Volkswagen_Golf_1.6_United  privat   
3     20160312  Smart_smart_fortwo_coupe_softouch/F1/Klima/Pan...  privat   
4     20160401  Ford_Focus_1_6_Benzin_TÜV_neu_ist_sehr_gepfleg...  privat   

   offertype  price   abtest   vehicle_type  reg_year    gearbox  powerps  \
0  Offer/Bid   5000  control            bus      2004     manual      158   
1  Offer/Bid   8500  control          sedan      1997  automatic      286   
2  Offer/Bid   8990     test          sedan      2009     manual      102   
3  Offer/Bid   4350  control      small car      2007  automatic       71   
4  Offer/Bid   1350     test  station wagon      2003     manual        0   

   ... odometer_km  reg_month  fuel_type       brand unrepair_dmg  \
0  ...      150000          3        lpg     peugeot           no   
1  ...      150000          6   gasoline         bmw           no   
2  ...       70000          7   gasoline  volkswagen           no   
3  ...       70000          6   gasoline       smart           no   
4  ...      150000          7   gasoline        ford           no   

            ad_created nrs_of_pictures  postal_code            last_seen  \
0  2016-03-26 00:00:00               0        79588  2016-04-06 06:45:54   
1  2016-04-04 00:00:00               0        71034  2016-04-06 14:45:08   
2  2016-03-26 00:00:00               0        35394  2016-04-06 20:15:37   
3  2016-03-12 00:00:00               0        33729  2016-03-15 03:16:28   
4  2016-04-01 00:00:00               0        39218  2016-04-01 14:38:50   

                                              detail  
0                               160 NAVTECH ON BOARD  
1                  4 4 Liter HAMANN UMBAU Mega Optik  
2                                         1.6 United  
3            fortwo coupe softouch/F1/Klima/Panorama  
4  1 6 Benzin TÜV neu ist sehr gepflegt.mit Klima...  

[5 rows x 21 columns]

### Analysis Next Steps:

* Find the most common brand/model combinations
* split the `odometer_km` into groups, and use aggregation to see if average prices follow any pattern based on mileage - done above
* how much cheaper are cars with damage than their non-damaged counterparts?


In [71]:
autos.columns

Index(['date_listing', 'name', 'seller', 'offertype', 'price', 'abtest',
       'vehicle_type', 'reg_year', 'gearbox', 'powerps', 'model',
       'odometer_km', 'reg_month', 'fuel_type', 'brand', 'unrepair_dmg',
       'ad_created', 'nrs_of_pictures', 'postal_code', 'last_seen', 'detail'],
      dtype='object')

In [72]:
print(autos['model'].head(10))

0      other
1        7er
2       golf
3     fortwo
4      focus
5    voyager
6       golf
7       golf
8      arosa
9     megane
Name: model, dtype: object


In [73]:
brand_count = autos['brand'].value_counts(normalize=True)
common_brand = brand_count[brand_count > .05].index
print(common_brand)

Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford'], dtype='object')


In [74]:
model_count = autos['model'].value_counts(normalize=True)
common_model = model_count[model_count > .05].index
print(model_count)
print(common_model)

golf          0.083326
other         0.075818
3er           0.058780
polo          0.036167
corsa         0.035785
                ...   
kappa         0.000045
b_max         0.000022
i3            0.000022
rangerover    0.000022
200           0.000022
Name: model, Length: 244, dtype: float64
Index(['golf', 'other', '3er'], dtype='object')


As we can see, the most common model is `golf`, which makes sense since `Volkswagen` is the most common brand, meaning `Volkswagen's model Golf` is the most common combination.    

In [76]:
non_damaged = autos[autos['unrepair_dmg'] == 'no']['price'].mean()
damaged = autos[autos['unrepair_dmg'] == 'yes']['price'].mean() 
diff = (non_damaged - damaged) / non_damaged * 100

print('Cars with no unrepaired damage are on average',round(diff, 2), '‰ more expensive.')


Cars with no unrepaired damage are on average 68.72 ‰ more expensive.


# End